figshare/figshare.py

```diff
@@ -290,3 +293,37 @@ def get_file_details(self, article_id, file_id):
         response = issue_request('GET', url,
                                  headers=self.get_headers(token=self.token))
         return response
+
+    def retrieve_files_from_article(self, article_id, outpath=''):
+        """ Retrieve files and save them locally.
+
+        By default, files will be stored in the current working directory
+        under a folder called figshare_<article_id> by default.
+        Specify <outpath> for: <outpath>/figshare_<article_id>
+
+        Parameters
+        ----------
+        article_id : str or int
+            Figshare article ID
+
+        """
+
+        # Handles missing trailing forward slash
+        if outpath != '':
+            if outpath[-1] != '/':
+                outpath += '/'
+
+        # Get list of files
+        file_list = self.list_files(article_id)
+
+        n_files = len(file_list)
+        if n_files != 0:
+            dir0 = outpath+"figshare_{0}/".format(article_id)
+            os.makedirs(dir0, exist_ok=True) # This might require Python >=3.2
+
+            for ii in range(n_files):
+               t_name = dir0 + file_list[ii]['name']
+               t_url  = file_list[ii]['download_url']
+
+               r = requests.get(t_url, allow_redirects=True)
```

Comment:
allow_redirects is default.

The builtin function `urllib.request.urlretrieve` would probably be superior here?

Reply:
Yes, I'm aware, but I find it more clear so readers don't have to look at the `requests` documentation.  I've noted this in an end-of-line comment

I'll have to think about `urllib`. I went with `requests` from the minimal perspective.  See also the top response to this:
https://stackoverflow.com/questions/2018026/what-are-the-differences-between-the-urllib-urllib2-urllib3-and-requests-modul

Reply:
`urllib.request.urlretrieve` has a major benefit that it doesn't require the entire file being loaded into memory before writing it to disk. This is especially important for large downloads like ones that might be hosted on figshare. The corresponding implementation with [requests](https://requests.readthedocs.io/en/master/user/quickstart/#raw-response-content) is more verbose and confusing. When the standard lib does exactly what you want, it's best to use it.

Reply:
Another consideration is whether we need enable authentication by way of the `get_headers` function. Or will articles with an ID always be accessible without authentication.

Reply:
> Another consideration is whether we need enable authentication by way of the `get_headers` function. Or will articles with an ID always be accessible without authentication.

For what my group is planning, everything is intended to be public.  We can explore this at a later point with confidential files.  I'm sure you know this, but one of the metadata that we can retrieve is whether a dataset is public.  So we can use that to notify the user that this is not yet supported.

Reply:
> `urllib.request.urlretrieve` has a major benefit that it doesn't require the entire file being loaded into memory before writing it to disk. This is especially important for large downloads like ones that might be hosted on figshare. The corresponding implementation with [requests](https://requests.readthedocs.io/en/master/user/quickstart/#raw-response-content) is more verbose and confusing. When the standard lib does exactly what you want, it's best to use it.

Large files will matter.  Let me dig into this and get back to you.

Reply:
urllib implemented and quickly tested.